File: mobile/lib/shared/models/event.dart
enum EventSource { google, apple, outlook, whatsapp, email, slack, manual }

enum EventStatus { pending, approved, rejected, confirmed }

class CalendarEvent {
  final String id;
  final String title;
  final DateTime startTime;
  final DateTime endTime;
  final String? location;
  final String? description;
  final List<String> attendees;
  final EventSource source;
  final EventStatus status;
  final double? confidence;
  final bool isAllDay;
  final String? meetingLink;
  final DateTime createdAt;

  const CalendarEvent({
    required this.id,
    required this.title,
    required this.startTime,
    required this.endTime,
    this.location,
    this.description,
    this.attendees = const [],
    required this.source,
    this.status = EventStatus.confirmed,
    this.confidence,
    this.isAllDay = false,
    this.meetingLink,
    required this.createdAt,
  });

  bool get isPending => status == EventStatus.pending;
  bool get isApproved => status == EventStatus.approved;
  Duration get duration => endTime.difference(startTime);

  CalendarEvent copyWith({
    String? id,
    String? title,
    DateTime? startTime,
    DateTime? endTime,
    String? location,
    String? description,
    List<String>? attendees,
    EventSource? source,
    EventStatus? status,
    double? confidence,
    bool? isAllDay,
    String? meetingLink,
    DateTime? createdAt,
  }) {
    return CalendarEvent(
      id: id ?? this.id,
      title: title ?? this.title,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      location: location ?? this.location,
      description: description ?? this.description,
      attendees: attendees ?? this.attendees,
      source: source ?? this.source,
      status: status ?? this.status,
      confidence: confidence ?? this.confidence,
      isAllDay: isAllDay ?? this.isAllDay,
      meetingLink: meetingLink ?? this.meetingLink,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'title': title,
    'startTime': startTime.toIso8601String(),
    'endTime': endTime.toIso8601String(),
    'location': location,
    'description': description,
    'attendees': attendees,
    'source': source.name,
    'status': status.name,
    'confidence': confidence,
    'isAllDay': isAllDay,
    'meetingLink': meetingLink,
    'createdAt': createdAt.toIso8601String(),
  };

  factory CalendarEvent.fromJson(Map<String, dynamic> json) => CalendarEvent(
    id: json['id'] as String,
    title: json['title'] as String,
    startTime: DateTime.parse(json['startTime'] as String),
    endTime: DateTime.parse(json['endTime'] as String),
    location: json['location'] as String?,
    description: json['description'] as String?,
    attendees: (json['attendees'] as List<dynamic>?)
            ?.map((e) => e as String)
            .toList() ??
        [],
    source: EventSource.values.firstWhere(
      (e) => e.name == json['source'],
      orElse: () => EventSource.manual,
    ),
    status: EventStatus.values.firstWhere(
      (e) => e.name == json['status'],
      orElse: () => EventStatus.confirmed,
    ),
    confidence: (json['confidence'] as num?)?.toDouble(),
    isAllDay: json['isAllDay'] as bool? ?? false,
    meetingLink: json['meetingLink'] as String?,
    createdAt: DateTime.parse(json['createdAt'] as String),
  );
}

File: backend/app/models/event.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, JSON, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Event(Base):
    __tablename__ = "events"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    start_time: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    end_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    location: Mapped[str | None] = mapped_column(String(500), nullable=True)
    # List of {name: str, email: str}
    attendees: Mapped[list] = mapped_column(JSON, default=list, nullable=False)
    # whatsapp / email / manual / calendar_sync
    source: Mapped[str] = mapped_column(String(50), default="manual", nullable=False)
    # 0.0 – 1.0
    confidence_score: Mapped[float | None] = mapped_column(Float, nullable=True)
    # pending / approved / rejected / synced
    status: Mapped[str] = mapped_column(String(50), default="pending", nullable=False, index=True)
    # Google Calendar / Outlook event ID after sync
    external_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    # Original text that triggered AI detection
    raw_text: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )

    user: Mapped["User"] = relationship("User", back_populates="events")  # noqa: F821
